src/vehicle_gui.cpp

```diff
@@ -3192,7 +3214,22 @@ struct VehicleViewWindow : Window {
 	{
 		if (str == nullptr) return;
 
-		Command<CMD_RENAME_VEHICLE>::Post(STR_ERROR_CAN_T_RENAME_TRAIN + Vehicle::Get(this->window_number)->type, this->window_number, str);
+		const Vehicle *v = Vehicle::Get(this->window_number);
+		switch (query) {
+			case QueryType::RenameVehicle:
+				Command<CMD_RENAME_VEHICLE>::Post(STR_ERROR_CAN_T_RENAME_TRAIN + v->type, v->index, str);
+				break;
+			case QueryType::RenameGroup:
+				if (v->group_id == DEFAULT_GROUP)
+					Command<CMD_ADD_VEHICLE_GROUP>::Post(STR_ERROR_GROUP_CAN_T_ADD_VEHICLE, NEW_GROUP, v->index, true /* add shared */);
+				if (v->group_id != DEFAULT_GROUP) // Not guaranteed if CMD_ADD_VEHICLE_GROUP fails.
```

Comment:
Can't be `else`, if adding to group failed it will still be `DEFAULT_GROUP`

Reply:
Actually... `else` is "fine", or rather... this bit of code has a major problem and multiplayer behaves as if there is an else.
In multiplayer `Post` is essentially asynchronous (and think of it as running several game ticks later), so if it makes a new group it won't rename it because the creation of the group has not yet happened when doing this check.

Reply:
In case it is not clear, this code will not work in multiplayer. Posting commands is asynchronous, and the command will not have been executed by the time you check v->group_id again.

Reply:
I’ve noticed Post accepts a callback to perform further processing after the command executes. If it accepts a custom state through a lambda, I’ll use that. Otherwise I’ll resort to adjusting the commands to perform both operations as a single command, or a adding a state somewhere in the vehicle window to store the new name.

Reply:
Command callbacks can have various argument formats. One of them allows you to get all the command parameters in addition to the command result. For example you can look at `CcCreateGroup`.

Reply:
@michicc That wouldn’t be enough. I need to store the result of ShowQueryString (i.e. the new group name) and reuse it in the callback. CMD_ADD_VEHICLE_GROUP does not take that kind of argument, which is why I’m thinking a lambda capture would be well-suited.